protocols/kad/src/behaviour/test.rs

```diff
@@ -223,3 +224,79 @@ fn unresponsive_not_returned_indirect() {
         }))
         .unwrap();
 }
+
+
+#[test]
+fn search_for_unknown_value() {
+    let (port_base, mut swarms) = build_nodes(3);
+
+    let peer_ids: Vec<_> = swarms.iter()
+        .map(|swarm| Swarm::local_peer_id(&swarm).clone()).collect();
+
+    swarms[0].add_address(&peer_ids[1], Protocol::Memory(port_base + 1).into());
+    swarms[1].add_address(&peer_ids[2], Protocol::Memory(port_base + 2).into());
+
+    let target_key = multihash::encode(Hash::SHA2256, &vec![1,2,3]).unwrap();
+    swarms[0].get_data(target_key);
+
+    Runtime::new().unwrap().block_on(
+        future::poll_fn(move || -> Result<_, io::Error> {
+            for swarm in &mut swarms {
+                loop {
+                    match swarm.poll().unwrap() {
+                        Async::Ready(Some(KademliaOut::GetValueRes { result, .. })) => {
+                            assert_eq!(result, None);
+                            return Ok(Async::Ready(()));
+                        }
+                        Async::Ready(_) => (),
+                        Async::NotReady => break,
+                    }
+                }
+            }
+
+            Ok(Async::NotReady)
+        }))
+        .unwrap()
+}
+
+#[test]
+fn put_value() {
+    let (port_base, mut swarms) = build_nodes(3);
+
+    let peer_ids: Vec<_> = swarms.iter()
+        .map(|swarm| Swarm::local_peer_id(&swarm).clone()).collect();
+
+    swarms[0].add_address(&peer_ids[1], Protocol::Memory(port_base + 1).into());
+    swarms[1].add_address(&peer_ids[2], Protocol::Memory(port_base + 2).into());
+
+    let target_key = multihash::encode(Hash::SHA2256, &vec![1,2,3]).unwrap();
+    swarms[0].get_data(target_key.clone());
+    swarms[1].put_data(target_key.clone(), &vec![4,5,6]);
+
+    Runtime::new().unwrap().block_on(
+        future::poll_fn(move || -> Result<_, io::Error> {
+            for swarm in &mut swarms {
+                loop {
+                    match swarm.poll().unwrap() {
```

Comment:
Is there not a potential race here where the `get_value` request may be processed by the other swarm before the `put_value`? In other words, isn't it necessary to first wait for the `put` to complete before issuing the `get`?

Reply:
That is true, but since the `PUT_VALUE` will probably have no response to it, I am puzzled with possible ways to actually wait for it.

Reply:
Even if there is no response to `PUT_VALUE` there should still be a `KademliaOut::PutValueResult` when the query is complete. When that event is emitted, it must be guaranteed that the local node has stored the value and at that point a `get_value` from another node for that key must always succeed. Currently the test may rely on the fact that `put_value` immediately puts the record into local storage, even before starting the query, but it is better not to make that assumption here.